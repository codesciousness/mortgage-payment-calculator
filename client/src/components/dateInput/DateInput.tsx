import React from 'react';
import { DatePicker } from '@mui/x-date-pickers/DatePicker';
import { AdapterDateFns } from '@mui/x-date-pickers/AdapterDateFns';
import { LocalizationProvider } from '@mui/x-date-pickers/LocalizationProvider';
import TextField from '@mui/material/TextField';
import Label from '../label/Label';

type DateInputProps = {
    name: string;
    value: Date | null;
    width?: number | string;
    onChange: (date: Date | null, keyboardInputValue?: string | undefined) => void;
};

const DateInput = ({ name, value, width, onChange }: DateInputProps): JSX.Element => {
    const today = new Date();
    const minDate = new Date(`${today.getMonth()}/31/${today.getFullYear()}`);

    return (
        <div>
            <LocalizationProvider dateAdapter={AdapterDateFns}>
                <Label name={name}/>
                <DatePicker
                    value={value}
                    minDate={minDate}
                    onChange={onChange}
                    inputFormat='MM/yyyy'
                    openTo='month'
                    views={['year', 'month']}
                    renderInput={(params) =>
                        <TextField
                            {...params}
                            size='small'
                            sx={{
                                width,
                                marginBottom: '0.75rem',
                                '& .MuiOutlinedInput-root': {
                                    backgroundColor: 'White',
                                    '& fieldset': {
                                        border: '1px solid Black',
                                    },
                                    '&.Mui-focused fieldset': {
                                        borderColor: 'DeepSkyBlue',
                                    }
                                }
                            }}
                        />
                    }
                />
            </LocalizationProvider>
        </div>
    );
};

export default DateInput;