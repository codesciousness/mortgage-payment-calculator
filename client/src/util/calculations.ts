export const isInt = (num: number | string): boolean => {
    let number: number;
    if (typeof num === 'string') number = stringToNum(num)
    else number = num;
    return Number.isInteger(number);
};

export const rmvCommas = (num: string): string => {
    return num.replaceAll(',', '');
};

export const addCommas = (num: number | string): string => {
    let numString: string;
    if (typeof num === 'number') numString = numToString(num)
    else numString = rmvCommas(num);
    return numString.replace(/\B(?<!\.\d*)(?=(\d{3})+(?!\d))/g, ",");
};

export const stringToNum = (str: string): number => {
    const num = parseFloat(rmvCommas(str));
    if (Number.isNaN(num)) return 0;
    return num;
};

export const numToString = (num: number): string => {
    return `${num}`;
};

export const round = (num: number | string): string => {
    let number: number;
    if (num === '') return '';
    if (typeof num === 'string') number = stringToNum(num)
    else number = num;
    return numToString(Math.round(number));
};

export const fixDecimal = (num: number | string): string => {
    let numString: string;
    let decimals: string;
    if (typeof num === 'number') numString = numToString(num)
    else numString = rmvCommas(num);
    decimals = numString.split('.')[1];
    if (decimals && decimals.length > 2) return numString.slice(0, numString.length - 1)
    else return numString;
};

export const formatAmount = (num: number | string): string => {
    return addCommas(round(num));
};

export const formatPercent = (num: number | string): string => {
    return addCommas(fixDecimal(num));
};

export const fromPercent = (percent: string, total: string): string => {
    const percentNum = stringToNum(percent)/100;
    const totalNum = stringToNum(total);
    const amount = percentNum * totalNum;
    return formatAmount(amount);
};

export const toPercent = (amount: string, total: string): string => {
    const amountNum = stringToNum(amount);
    const totalNum = stringToNum(total);
    const percent = (amountNum / totalNum) * 100;
    if (isInt(percent)) {
        return addCommas(percent);
    }
    return addCommas(percent.toFixed(2));
};

export const mortgageFormula = (loanAmount: string, interestRate: number, loanTerm: number): string => {
    const principal = stringToNum(loanAmount);
    const r = interestRate/12;
    const n = loanTerm * 12
    const monthlyPayment = principal * (r*(1 + r)**n)/((1 + r**n) - 1);
    return numToString(monthlyPayment);
};