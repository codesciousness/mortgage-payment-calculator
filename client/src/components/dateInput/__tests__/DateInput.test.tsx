import React from 'react';
import { render, screen } from '@testing-library/react';
import { create } from 'react-test-renderer';
import DateInput from '../DateInput';
import { dateToString } from '../../../util/calculations';

const name = 'Label Name';
const width = 300;
const onChange = () => null;

describe('<DateInput />', () => {
    test('DateInput component is defined', () => {
        expect(<DateInput name={name} value={null} onChange={onChange}/>).toBeDefined();
    });
    test('component matches the snapshot', () => {
        const component = create(<DateInput name={name} value={new Date()} width={width} onChange={onChange}/>);
        expect(component.toJSON()).toMatchSnapshot();
    });
    test('renders correctly when value prop is a date', () => {
        render(<DateInput name={name} value={new Date()} width={width} onChange={onChange}/>);
        expect(screen.getByText('Label Name')).toBeInTheDocument();
        expect(screen.getByDisplayValue(dateToString(new Date()))).toBeInTheDocument();
    });
    test('renders correctly when value prop is null', () => {
        render(<DateInput name={name} value={null} width={width} onChange={onChange}/>);
        expect(screen.getByText('Label Name')).toBeInTheDocument();
        expect(screen.queryByDisplayValue(dateToString(new Date()))).not.toBeInTheDocument();
    });
});