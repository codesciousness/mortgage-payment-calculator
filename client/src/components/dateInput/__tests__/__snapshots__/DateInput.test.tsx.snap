// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DateInput /> component matches the snapshot 1`] = `
<div>
  <label
    className="MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-root MuiFormLabel-colorPrimary css-22ecsy-MuiFormLabel-root-MuiInputLabel-root"
    id="LabelName"
  >
    Label Name
  </label>
  <div
    className="MuiFormControl-root MuiTextField-root css-dmaw38-MuiFormControl-root-MuiTextField-root"
  >
    <div
      className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        aria-label="Choose date, selected date is May 30, 2023"
        aria-readonly={true}
        autoFocus={false}
        className="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
        id=":r0:"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        readOnly={true}
        required={false}
        type="text"
        value="05/2023"
      />
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-hdw1oc"
        >
          <span
            className="notranslate"
          >
            â€‹
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;
